Source/JavaScriptCore/wasm/WasmBBQJIT.cpp

```diff
@@ -4139,10 +4184,21 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addCallIndirect(unsigned tableIndex, co
             m_jit.loadPtr(Address(calleeSignatureIndex, FuncRefTable::Function::offsetOfValue()), jsCalleeAnchor);
 #endif
             ASSERT(static_cast<ptrdiff_t>(WasmToWasmImportableFunction::offsetOfSignatureIndex() + sizeof(void*)) == WasmToWasmImportableFunction::offsetOfEntrypointLoadLocation());
+
+            // Save the table entry in calleeRTT in case we need it for the GC check below.
+            if (Options::useWebAssemblyGC())
+                m_jit.move(calleeSignatureIndex, calleeRTT);
             m_jit.loadPairPtr(calleeSignatureIndex, TrustedImm32(FuncRefTable::Function::offsetOfFunction() + WasmToWasmImportableFunction::offsetOfSignatureIndex()), calleeSignatureIndex, calleeCode);
 
             throwExceptionIf(ExceptionType::NullTableEntry, m_jit.branchTestPtr(ResultCondition::Zero, calleeSignatureIndex, calleeSignatureIndex));
-            throwExceptionIf(ExceptionType::BadSignature, m_jit.branchPtr(RelationalCondition::NotEqual, calleeSignatureIndex, TrustedImmPtr(TypeInformation::get(originalSignature))));
+            auto indexEqual = m_jit.branchPtr(CCallHelpers::Equal, calleeSignatureIndex, TrustedImmPtr(TypeInformation::get(originalSignature)));
+
+            if (Options::useWebAssemblyGC())
+                addRTTSlowPathJump(originalSignature.index(), calleeRTT);
+            else
+                emitThrowException(ExceptionType::BadSignature);
+
+            indexEqual.link(&m_jit);
```

Comment:
Is it possible to avoid doing sub typing check when it is unnecessary? `call_indirect` is super important path, so we cannot regress performance for the existing use. I wonder if we always need to have this check for all the functions, like, for example, do we need this even for `i32 -> i32` function?

Reply:
I thought about this a bit, and there is one optimization that's missing here that I think will address your concern in most code.

In the GC proposal there are final and non-final `sub` types. I believe old non-GC code types will all be considered final by default. For a final target type we can avoid emitting the check entirely since it cannot have any subtypes by definition.

Reply:
The patch now only emits the extra checks in BBQ and OMG when the target type is non-final, which is only the case for new GC types.